common/src/main/java/org/apache/celeborn/common/protocol/PartitionLocation.java

```diff
@@ -281,12 +365,26 @@ public boolean equals(Object other) {
         && host.equals(o.host)
         && rpcPort == o.rpcPort
         && pushPort == o.pushPort
-        && fetchPort == o.fetchPort;
+        && fetchPort == o.fetchPort
+        && internalPort == o.internalPort
```

Comment:
Interesting idea.
QQ: currently the 'identity' is tied to host/ports - and so shipped around in worker info and/or partition location.
Is there an indirection to this ? (in spark, for example, executor id - which encapsulates a host/port for example).

If yes, we can put all the host/ports behind that - and have only the identity propagated as part of the messages ? Thoughts ?

Reply:
> I think we can reuse rpcPort/pushPort/fetchPort instead of introducing secure companions

@waitinfuture So the reason I introduced these ports was because in our last discussion, we reached the conclusion that we will have different ports for secured connection than unsecured communication. Otherwise, that will be backward compatibility issues between clients that are still on older version and servers which have been upgraded.

You are right about `internalPort`. It can be removed from `PartitionLocation`. Since, I had to add other ports, I added that as well.

> Is it necessary to put internalPort/securedRpcPort/securedPushPort/securedFetchPort inside PartitionLocation? This can cause backward compatibility issue when rolling upgrade where previously auth is disabled but after upgrading auth is enabled (same to WorkerInfo). 

About this, my understanding is that we can achieve this without any issues is by first upgrading the binaries of both servers and clients without turning on authentication on either side. Once, this is done, then auth can be turned on the servers. 

I am also open to other suggestions as @mridulm pointed out.

Reply:
@mridulm Unfortunately there is no indirection to PartitionLocation or WorkerInfo for now. As you said, Worker is identified by host/ports, and PartitionLocation is identified by uniqueId plus WorkerInfo. This has worked well for long. We may not want to add a UUID to PartitionLocation since the message is quite big already when there are lots of PartitionLocations.

@otterc For the backward compatibility, I think you are right about PartitionLocations, but for WorkerInfos, the restarted new Workers have secured ports that are not equal to zero, so when it registers itself to Master, Master will think it's a new Worker, this may lead to issues.

> So the reason I introduced these ports was because in our last discussion
I totally agree to introduce these secure ports. What I mean is that we don't need to add new variables in PartitionLocation, instead we can reuse the existing variables (rpcPort/pushPort/fetchPort) to store the actual secured ports if auth is enabled.

Reply:
> but for WorkerInfos, the restarted new Workers have secured ports that are not equal to zero, so when it registers itself to Master, Master will think it's a new Worker, this may lead to issues.

Is this because Master maintains some state in Ratis for each worker? I would think this would be a problem even now because by default we are using ephemeral ports for Workers. Every time the worker is restarted, it will use different ports for RPC, push, and fetch. Let me know if I am missing something.

> What I mean is that we don't need to add new variables in PartitionLocation, instead we can reuse the existing variables (rpcPort/pushPort/fetchPort) to store the actual secured ports if auth is enabled.

Got it. For this, the servers would then have to maintain a state whether an app is secured or not. I think this is needed anyways if we want to prevent an app that securely talks to Master but tries to communicate with the workers on the unsecured port (CELEBORN-1261). I can go ahead with this. WDYT @mridulm?

Reply:
> Every time the worker is restarted, it will use different ports for RPC, push, and fetch. 

This is in fact fine because workers with the same ip but different ports are considered different workers.

Reply:
There are a few combinations here we need to ensure work fine - for example, some subset of apps which are secured while others are not, etc @otterc

Reply:
@waitinfuture Interesting, after the authn work, let us relook at whether we can make the existing message sizes more compact if possible. Thanks for the details.

Reply:
> There are a few combinations here we need to ensure work fine - for example, some subset of apps which are secured while others are not

I think removing the changes to `PartitionLocation` from this PR and then addressing it in a separate JIRA might be better? Sending `PartitionLocations` to client based on whether the app is secured or unsecured, maybe more complex. We can decide which approach to take in a followup change. WDYT @mridulm @waitinfuture

Reply:
Sounds good to me.

Reply:
> Why would this be an issue if the worker is restarted with authentication enabled? The earlier instance stops and de-registers itself, and then a new worker instance comes up with different ports. Why does the Master need to associate it with the previous instance?

This relates to graceful shutdown. If `celeborn.worker.graceful.shutdown.enabled` is enabled and workers are configured with fixed ports, when restart Worker, it will save its in-memory states(mostly PartitionLocations) in local leveldb(or rocksdb) and restore after restart, so it can serve shuffle read normally without interfering the application. This feature requires that workers are configured with fixed ports. If the Master thinks it is a new worker, after `celeborn.master.heartbeat.worker.timeout` the Master will mark the previous workers as lost and return to LifecycleManager through `HeartbeatFromApplicationResponse` and propagate to ShuffleClient, this may cause an issue.

Reply:
More on rolling upgrade and graceful shutdown here https://celeborn.apache.org/docs/latest/upgrading/

Reply:
Ok. Then we can incorporate this logic for `WorkerInfos` equality:

```
if (celeborn.worker.graceful.shutdown.enabled  &&  !disable_unsecured_ports) 
       then WorkerInfos equality should exclude the secured ports
else if (disable_unsecured_ports) 
       then WorkerInfos equality should exclude the unsecured ports
else 
      include all the ports.
```
Note: disable_unsecured_ports config is not added yet.
cc. @waitinfuture @mridulm

Reply:
Sounds good to me, thanks! @otterc